src/dbnode/persist/fs/msgpack/encoder_fast.go

```diff
@@ -0,0 +1,348 @@
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE
+
+package msgpack
+
+import (
+	"math"
+
+	"github.com/m3db/m3/src/dbnode/persist/schema"
+
+	"gopkg.in/vmihailenco/msgpack.v2/codes"
+)
+
+// EncodeLogEntryFast encodes a commit log entry with no buffering and using optimized helper
+// functions that bypass the msgpack encoding library by manually inlining the equivalent code.
+//
+// The reason we had to bypass the msgpack encoding library is that during perf testing we found that
+// this function was spending most of its time setting up stack frames for function calls. While
+// the overhead of a function call in Golang is small, when every helper function does nothing more
+// than write a few bytes to an in-memory array the function call overhead begins to dominate,
+// especially when each call to this function results in dozens of such helper function calls.
+//
+// Manually inlining the msgpack encoding results in a lot of code duplication for this one path, but
+// we pay the price because this is the most frequently called function in M3DB and it indirectly
+// applies back-pressure on every other part of the system via the commitlog queue. As a result, almost
+// any performance gains that can be had in this function are worth it.
+//
+// Before modifying this function, please run the BenchmarkLogEntryEncoderFast benchmark as a small
+// degradation in this functions performance can have a substantial impact on M3DB.
+func EncodeLogEntryFast(b []byte, entry schema.LogEntry) ([]byte, error) {
+	if logEntryHeaderErr != nil {
+		return nil, logEntryHeaderErr
+	}
+
+	b = append(b, logEntryHeader...)
+	b = encodeVarUint64(b, entry.Index)
+	b = encodeVarInt64(b, entry.Create)
+	b = encodeBytes(b, entry.Metadata)
+	b = encodeVarInt64(b, entry.Timestamp)
+	b = encodeFloat64(b, entry.Value)
+	b = encodeVarUint64(b, uint64(entry.Unit))
+	b = encodeBytes(b, entry.Annotation)
+
+	return b, nil
+}
+
+// encodeVarUint64 is used in EncodeLogEntryFast which is a very hot path.
+// As a result, many of the function calls in this function have been
+// manually inlined to reduce function call overhead.
```

Comment:
So I’ve read this code a few times, the manual inlining makes a lot of sense.  Having said that it might be worth using Genny to code gen the code so that if we really need to change parts of the code it affects every party (DRY it up).  

Genny should be able to code gen all the inlined parts, it’s more work for sure, so I’m fine if with TODO it but we should definitely do this at some point (otherwise the copy paste manual inlining thing may become a pattern and code gen is definitely better than what manual copy paste to inline things).

Thoughts?

Reply:
Yeah I thought about doing that too, but I’d prefer to leave as a TODO for now. Mainly because:

1. Hopefully this doesn’t become a pattern for us
2. This is really unplanned work and there’s like 5 other things I’m supposed to be doing that are not performance improvements and I have a feeling getting the genny thing right will take a little bit of back and forth

Reply:
This PR makes sense only if you believe Go won't inline your function in the near future. Try debugging this the explanation here in the Go compiler: https://github.com/golang/go/blob/master/src/cmd/compile/internal/gc/inl.go#L5-L25.

Reply:
I think the benchmarks demonstrate that this P.R is worth it. Even if Golang inlining gets better in the near-future, this function is so hot and the number of function calls we removed so high (we're talking about literally dozens of function calls per iteration that have been removed) I highly doubt that the Go compiler will be able to achieve a similar effect anytime soon